File: ThreatModelDfdService/Services/Impl/DfdService.cs
using ThreatModelDfdService.Data.DTO;
using ThreatModelDfdService.Model.Context;
using ThreatModelDfdService.Model.Entity;

namespace ThreatModelDfdService.Services.Impl;

public class DfdService(
    DfdElementService dfdElementService,
    DataFlowService dataFlowService,
    MSSQLContext context)
{
    public async Task<FullDfdResponseDTO> SyncElementsAsync(long dfdId, UpsertAllDfdElementsDTO dto)
    {
        foreach (var elementDto in dto.Elements)
        {
            await dfdElementService.CreateOrUpdateAsync(dfdId, elementDto);
        }

        await context.SaveChangesAsync();

        foreach (var dataFlowDto in dto.DataFlows)
        {
            await dataFlowService.CreateOrUpdateDataFlow(dataFlowDto, dfdId);
        }
        await context.SaveChangesAsync();
        
        return new FullDfdResponseDTO(
            dfdElementService.GetDfdElementsByDfdId(dfdId),
            dataFlowService.GetDataFlowsByDfdId(dfdId)
        );
    }

    public DfdDTO CreateChildDfd(CreateDfdChildDTO dto)
    {
        DfdElement processParent = dfdElementService.GetById(dto.ProcessParentId);
        Dfd dfd = FindById(processParent.DfdId);
        Dfd childDfd = Create(dto.LevelNumber + 1, dfd.ProjectId, processParent.DfdId);

        Process process = (Process) processParent;
        process.DfdChildId = childDfd.Id;
        context.SaveChanges();

        return new DfdDTO(childDfd.Id, childDfd.DfdParentId, childDfd.LevelNumber, []);
    }

    public Dfd Create(int LevelNumber, long projectId, long? dfdParentId = null)
    {
        Dfd dfd = context.Dfds.Add(new Dfd {
            LevelNumber = LevelNumber,
            ProjectId = projectId,
            DfdParentId = dfdParentId
        }).Entity;
        context.SaveChanges();
        return dfd;
    }

    public DfdDTO GetDfdById(long id)
    {
        Dfd dfd = FindById(id);
        return new DfdDTO(dfd.Id, dfd.DfdParentId, dfd.LevelNumber, dfdElementService.GetDfdElementsByDfdId(id));
    }

    public Dfd FindById(long id)
    {
        Dfd? dfd = context.Dfds.Find(id);
        if (dfd == null) throw new KeyNotFoundException("DFD not found with the provided ID: " + id);
        return dfd;
    }
}

File: dfd-service/src/main/java/com/tcc/dfd_service/service/DfdService.java
package com.tcc.dfd_service.service;

import com.tcc.dfd_service.dto.*;
import com.tcc.dfd_service.entity.Dfd;
import com.tcc.dfd_service.entity.DfdElement;
import com.tcc.dfd_service.entity.Process;
import com.tcc.dfd_service.repository.DfdRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class DfdService {

    private final DfdRepository dfdRepository;
    private final DfdElementService dfdElementService;
    private final DataFlowService dataFlowService;

    @Transactional
    public FullDfdResponseDTO syncElementsAndFlows(Long dfdId, SyncDfdElementsAndFlowsDTO dto) {

        for (UpsertDfdElementDTO elementDto : dto.elements()) {
            dfdElementService.createOrUpdate(dfdId, elementDto);
        }

        for (DataFlowRequestDTO flowDto : dto.dataFlows()) {
            dataFlowService.createOrUpdateDataFlow(flowDto, dfdId);
        }

        return new FullDfdResponseDTO(
                dfdElementService.getDfdElementsByDfdId(dfdId),
                dataFlowService.getDataFlowsByDfdId(dfdId)
        );
    }

    @Transactional
    public DfdDTO createChildDfd(CreateDfdChildDTO dto) {
        DfdElement processParent = dfdElementService.getById(dto.processParentId());

        Dfd parentDfd = findById(processParent.getDfdId());

        Dfd childDfd = create(
                        dto.levelNumber() + 1,
                        parentDfd.getProjectId(),
                        parentDfd.getId()
                );

        Process process = (Process) processParent;

        process.setDfdChild(childDfd);

        return new DfdDTO(
                childDfd.getId(),
                childDfd.getDfdParent() != null
                        ? childDfd.getDfdParent().getId()
                        : null,
                childDfd.getLevelNumber(),
                List.of(),
                List.of()
        );
    }

    @Transactional
    public Dfd create(Integer levelNumber, Long projectId, Long dfdParentId) {

        Dfd dfd = new Dfd();
        dfd.setLevelNumber(levelNumber);
        dfd.setProjectId(projectId);

        if (dfdParentId != null) {
            dfd.setDfdParent(findById(dfdParentId));
        }

        return dfdRepository.save(dfd);
    }

    public DfdDTO getDfdById(Long id) {
        Dfd dfd = findById(id);

        return new DfdDTO(
                dfd.getId(),
                dfd.getDfdParent() != null
                        ? dfd.getDfdParent().getId()
                        : null,
                dfd.getLevelNumber(),
                dfdElementService.getDfdElementsByDfdId(id),
                dataFlowService.getDataFlowsByDfdId(id)
        );
    }

    public Dfd findById(Long id) {
        return dfdRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("DFD not found with the provided ID: " + id));
    }
}